### ISSUE TITLE
Chapter 10 Notes, Errata, and Discussion

### ISSUE THREAD

**Issue by lihaoyi** [handsonscala/handsonscala#10]
Thread to discuss anything and everything about Chapter 10. Important topics will be folded into the issue description.

**Comment by hansdejong** [comment]
I don't understand why I can't copy the project.
I had a directory Ch10.4, copying to Ch10.5 failed. I had to copy post and build.sc first and then download the mill-launcherscript (even after a computer-restart). What is the nature of the connection between the project and the "mill-worker"?
(Btw., enjoying the book!)

**Comment by hansdejong** [comment]
Is it possible to pretty-print the html?
I see the ammonite-REPL knows how to pretty-print a "scalatags.Text.TypedTag[String]", but toString(), render(), writeTo(java.io.Writer) don't. What is the REPL's secret?

**Comment by hansdejong** [comment]
Is there a setup-example for a Scala.js/jvm cross-compilation mill project, including shared code?

**Comment by lihaoyi** [comment]
@hansdejong Scala.js is beyond the scope of this book, but you can look https://github.com/lihaoyi/fansi for an example of a cross-compiled module where all the code is shared (platform-specific code would go in `src-jvm/` and `src-js/` folders)

**Comment by ooraini** [comment]
I am trying to do the following:

    val mains = T.input {`spark-jobs`.discoverMainClasses()}
    object `spark-job` extends Cross[SparkJobModule](mains():_*)
    class SparkJobModule(main: String) extends Module {
    }

where spark-jobs is a Scala module, I know that you can't modify the target graph based on the result of some target, we used `interp.watch` in the book, where changes on the file system triggered rebuilding the graph, but is their a work around for the general case?

**Comment by lihaoyi** [comment]
@omaraloraini nope, `interp.watch` is the most flexible option you get. The inability to re-configure the graph half way through evaluation is intentional, and necessary for a lot of how Mill currently works

**Comment by staniopolis** [comment]
Any ideas?
build.sc identical to example
```
stas@Stanislau:/mnt/d/HandsOnScala/BLOG$ ./mill show dist
Failed to resolve ivy dependencies:No such file or directory
```

**Comment by amnaredo** [comment]
I'm getting this error too:

```
$ ./mill show dist
Failed to resolve ivy dependencies:Boxed Exception
```

**Comment by jk-1** [comment]
These may be strange questions, but I really try to understand how the mill is working. This is about mill v0.9.9.

https://com-lihaoyi.github.io/mill/mill/Tasks.html

In Commands chapter it is said:

> Commands can be parametrized, but their output is not cached, so they will re-evaluate every time even if none of their inputs have changed.

I have a simple hello world app in simple mod01 ScalaModule and after `mill show mod01.run the` `meta.json` file is as follows:

    cat out/mod01/run/meta.json 
    {
        "value": {
            
        },
        "valueHash": 0,
        "inputsHash": -1701922531
    }


Does the "output is not cached" cause the empty value and valueHash = 0 in the json?

What is the inputsHash -part in json file? Is it used to detect if re-eval is needed? If yes then why is that needed in T.command because "they will re-evaluate every time even if none of their inputs have changed"?

**Comment by cdcd88** [comment]
I am on section 10.4.3 and finished refactored `build.sc`. From my understanding, Mill generates new files only when there is a change. However, I found that it generates all of them even if I make a change to one file.

For example, I change `1 - My First Post.md` then run `mill dist`. After that, I go to `out/post/2/render` or `out/post/3/render`, run `ls -al` and the date changed to the same minute I ran `mill dist`.

Why does Mill regenerate the two other files I didn't touch?

Here is the code I have on my machine
```scala
import $ivy.`com.lihaoyi::scalatags:0.9.1`, scalatags.Text.all._
import $ivy.`com.atlassian.commonmark:commonmark:0.13.1`
import mill._

interp.watch(os.pwd / "post")

// _1: 1, 2, 3
// _2: title
// _3: path
val postInfo = os
  .list(os.pwd / "post")
  .map{ p =>
    val s"$prefix - $suffix.md" = p.last
    (prefix, suffix, p)
  }
  .sortBy(_._1.toInt)

val bootstrapCss = link(
  rel := "stylesheet",
  href := "https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.css"
)

def mdNameToHtml(name: String) = name.replace(" ", "-").toLowerCase + ".html"

object post extends Cross[PostModule](postInfo.map(_._1):_*)
class PostModule(number: String) extends Module {
  val Some((_, suffix, markdownPath)) = postInfo.find(_._1 == number)
  def path = T.source(markdownPath)
  def render = T{
    val parser = org.commonmark.parser.Parser.builder().build()
    val document = parser.parse(os.read(path().path))
    val renderer = org.commonmark.renderer.html.HtmlRenderer.builder().build()
    val output = renderer.render(document)
    os.write(
      T.dest / mdNameToHtml(suffix),
      doctype("html")(
        html(
          head(bootstrapCss),
          body(
            h1(a(href := "../index.html")("Blog"), " / ", suffix),
            raw(output),
          )
        )
      )
    )
    PathRef(T.dest / mdNameToHtml(suffix))
  }
}

def links = T.input{ postInfo.map(_._2) }

def index = T{
  os.write(
    T.dest / "index.html",
    doctype("html")(
      html(
        head(bootstrapCss),
        body(
          h1("Blog"),
          for(suffix <- links())
          yield h2(a(href := ("post/" + mdNameToHtml(suffix)))(suffix))
        )
      )
    )
  )
  PathRef(T.dest / "index.html")
}

val posts = T.sequence(postInfo.map(_._1).map(post(_).render))

def dist = T {
  for (post <- posts()) {
    os.copy(post.path, T.dest / "post" / post.path.last, createFolders = true)
  }
  os.copy(index().path, T.dest / "index.html")

  PathRef(T.dest)
}
```
